### ISSUE TITLE
`WithPgAdmin` does not support `IResourceBuilder<PostgresServerResource>`

### ISSUE THREAD

**Issue by mitchdenny** [dotnet/aspire#1532]
We need to make it work for this as well. It also doesn't appear to be pre-configuring the server list correctly, although that might just be a glitch locally.

**Comment by andrevlins** [comment]
Hi @mitchdenny,

In order for ```withPgAdmin``` to be able to pre-configure the postgresql server, the port must be defined by the developer.

```csharp
var catalogDb = builder.AddPostgresContainer("postgres", port: 5234, password: "123Postgres")
                    .WithPgAdmin(hostPort: 8081)
                    .AddDatabase("catalogdb");
```
This limitation is due to the impossibility of knowing the dynamically allocated ports before starting the pgAdmin4 container. Even after adding the new lifecycle ```AfterEndpointsAllocatedAsync``` it was not possible because at the point this method is called it is not possible to add new containers to the application model.

So the current implementation will ignore the server preconfiguration if the port is dynamically allocated.

That said, we are also limited to only supporting ```IResourceBuilder<PostgresContainerResource>``` as it allows us to specify the port through ```AddPostgresContainer```.

I could support ```IResourceBuilder<PostgresServerResource>``` as long as it's okay to add the port as an optional parameter to ```AddPostgres```. 

What do you think?

Another issue taking advantage of the interaction is that I no longer see the justification for making the hostPort parameter mandatory for ```withPgAdmin``` this was really a mistake. Should I open an issue + PR to address this or just the PR?

**Comment by mitchdenny** [comment]
I've got a PR where we make this work. In Aspire we have these things called lifecycle hooks which can be registered to run and mutate the app model after the endpoints have been allocated. Check out:

https://github.com/dotnet/aspire/pull/1552

I've also made it work with PostgresServerResource and PostgresContainerResource with this change. It also pre-configures the passwords as well so you don't need to cut paste them.

**Comment by andrevlins** [comment]
I just saw your PR using lifecycle to defer preconfiguration of servers. This approach of modifying the file mounted in ```AfterEndpointsAllocatedAsync``` was what I was missing!

Glad you fixed it. It looks great.

Can you take advantage of this PR and adjust the issue of the requirement to pass the hostport to ```withPgAdmin``` ?

**Comment by mitchdenny** [comment]
Yeah I just fixed that up in the latest push.
